docs/custom-components.mdx

```diff
@@ -0,0 +1,286 @@
+
+```bash
+# Build the component
+make build
+
+# Test locally
+make wasi-exec
+```
+
+### `wavs-wasi-chain` crate
+
+The `wavs-wasi-chain` crate provides a set of helpful functions for making HTTP requests and interacting with the blockchain. It also provides a macro for decoding trigger data for use in the component.
+
+Learn more in the [crate documentation](https://docs.rs/wavs-wasi-chain/latest/wavs_wasi_chain/all.html#functions).
+
+### Network requests
+
+Components can make network requests to external APIs using the `wavs-wasi-chain` crate.
+
+The following dependencies are useful for making HTTP requests from a component. These are added to a component's `Cargo.toml` file:
+
+```toml Cargo.toml
+[dependencies]
+wavs-wasi-chain = { workspace = true }  # HTTP utilities
+wstd = { workspace = true }             # Runtime utilities
+serde = { workspace = true }            # Serialization
+serde_json = { workspace = true }       # JSON handling
+```
+
+The following example shows how to make a basic HTTP GET request from a component:
+
+```rust lib.rs
+async fn make_request() -> Result<YourResponseType, String> {
+    // Create the request
+    let url = "https://api.example.com/endpoint";
+    let mut req = http_request_get(&url).map_err(|e| e.to_string())?;
+    
+    // Add headers
+    req.headers_mut().insert(
+        "Accept",
+        HeaderValue::from_static("application/json")
+    );
+    
+    // Make the request and parse JSON response
+    let json: YourResponseType = fetch_json(req)
+        .await
+        .map_err(|e| e.to_string())?;
+        
+    Ok(json)
+}
+```
+
+Other functions are available in the [crate documentation](https://docs.rs/wavs-wasi-chain/latest/wavs_wasi_chain/all.html#functions).
+
+### Logging
+
+Components can use logging to debug and track the execution of the component.
+
+**Logging in development**:
+
+Use `println!()` to write to stdout/stderr. This is visible when running `wasi-exec` locally.
+
+```rust lib.rs
+println!("Debug message: {:?}", data);
+```
+
+**Logging in production**
+
+For production, you can use a `host::log()` function which takes a `LogLevel` and writes its output via the tracing mechanism. Along with the string that the developer provides, it attaches additional context such as the `ServiceID`, `WorkflowID`, and component `Digest`.
+
+```rust lib.rs
+host::log(LogLevel::Info, "Production logging message");
+```
+
+### Blockchain interactions
+
+Components can query Ethereum using the `wavs-wasi-chain` crate:
+
+```rust lib.rs
+use wavs_wasi_chain::host;  // For blockchain interactions
+use alloy_sol_types::SolValue;  // For ABI encoding/decoding
+
+// Get chain configuration
+let chain_config = host::get_eth_chain_config("ethereum")
+    .ok_or("Chain config not found")?;
+
+// Make an eth_call
+let result = host::eth_call(
+    &chain_config,
+    "0x...", // contract address
+    "0x...", // encoded function call
+    "latest" // block tag
+).await?;
+```
```

Comment:
where was this example pulled from?

Reply:
I'm not aware of our roadmap to add support for this. We did discuss it, and saw it as a cool stopgap before WASI p3 is released (where we can dynamic-link components and offer super cool helpers)- but iirc the decision was to _not_ do this, instead we offer [some basic helpers to create an alloy client](https://docs.rs/wavs-wasi-chain/0.3.0/wavs_wasi_chain/ethereum/fn.new_eth_provider.html) which need to be compiled into the component

We also offer host bindings to get the chain config for a given chain name in the config.toml: https://github.com/Lay3rLabs/wavs-wasi/blob/13701bfdc508071fb89444634aca664ecb7de68e/wit/wavs_worker.wit#L128

But we do not offer host bindings to talk to the chain